From 8268941ebc0053b9c577fc58a36249f9797eab93 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 16 Dec 2024 19:12:28 +0100
Subject: [PATCH 53/86] crypt32: Avoid stack-buffer-overflow in
 CryptSignMessage.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M crypt32.dll -p dlls/crypt32/tests/x86_64-windows/crypt32_test.exe message && touch dlls/crypt32/tests/x86_64-windows/message.ok
=================================================================
==632==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffffe1ff5f0 at pc 0x6ffff824d3b9 bp 0x7ffffe1fef80 sp 0x7ffffe1fefc8
READ of size 4 at 0x7ffffe1ff5f0 thread T0
027c:fixme:dbghelp:elf_search_auxv can't find symbol in module
027c:fixme:dbghelp_dwarf:dwarf2_load_one_entry Unhandled Tag type 0xa at debug_info(abbrev:00007E8F33C20B80,symt:0000000000000000) in ctx(00007E8F34E22200,L"crypt32")
027c:fixme:dbghelp_dwarf:dwarf2_load_one_entry Unhandled Tag type 0xa at debug_info(abbrev:00007E8F33D90B80,symt:0000000000000000) in ctx(00007E8F34F818E0,L"crypt32_test")
    #0 0x6ffff824d3b8 in CryptMsgOpenToEncode Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\crypt32\msg.c
    #1 0x6ffff82473d0 in CryptSignMessage Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\crypt32\message.c:491
    #2 0x0001400e11ad in func_message+0x728d (Z:\home\bernhard\data\entwicklung\2024\wine\wine-build\build-asan-pe\64\obj\dlls\crypt32\tests\x86_64-windows\crypt32_test.exe+0x1400e11ad)
    #3 0x00014015ecb1 in main+0x541 (Z:\home\bernhard\data\entwicklung\2024\wine\wine-build\build-asan-pe\64\obj\dlls\crypt32\tests\x86_64-windows\crypt32_test.exe+0x14015ecb1)
    #4 0x000140160bbf in mainCRTStartup Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\msvcrt\crt_main.c:58
    #5 0x6ffffbdc4808 in BaseThreadInitThunk Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\kernel32\thread.c:61
    #6 0x6ffffad1fa1a  (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1ff5f0 is located in stack of thread T0 at offset 272 in frame
    #0 0x6ffff8246c2f in CryptSignMessage Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\crypt32\message.c:409

  This frame has 5 object(s):
    [32, 40) 'hCryptProv' (line 410)
    [64, 68) 'freeProv' (line 411)
    [80, 84) 'keySpec' (line 412)
    [96, 144) 'signInfo' (line 415)
    [176, 272) 'signer' (line 416) <== Memory access at offset 272 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow Z:\home\bernhard\data\entwicklung\2024\wine\wine\dlls\crypt32\msg.c in CryptMsgOpenToEncode
Shadow bytes around the buggy address:
  0x7ffffe1ff300: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1ff380: f8 f8 f8 f8 f3 f3 f3 f3 f3 f3 f3 f3 00 00 00 00
  0x7ffffe1ff400: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ff480: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x7ffffe1ff500: 00 f2 f2 f2 04 f2 04 f2 00 00 00 00 00 00 f2 f2
=>0x7ffffe1ff580: f2 f2 00 00 00 00 00 00 00 00 00 00 00 00[f3]f3
  0x7ffffe1ff600: f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ff680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ff700: f1 f1 f1 f1 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2
  0x7ffffe1ff780: f8 f8 f2 f2 f8 f2 f2 f2 f8 f2 00 00 00 00 00 00
  0x7ffffe1ff800: 00 00 00 00 00 00 00 00 00 f2 f2 f2 f2 f2 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==632==ABORTING
make: *** [Makefile:36953: dlls/crypt32/tests/x86_64-windows/message.ok] Fehler 1
---
 dlls/crypt32/msg.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/crypt32/msg.c b/dlls/crypt32/msg.c
index 5e6d59a1f58..4cbdf5391b3 100644
--- a/dlls/crypt32/msg.c
+++ b/dlls/crypt32/msg.c
@@ -1424,8 +1424,8 @@ static HCRYPTMSG CSignedEncodeMsg_Open(DWORD dwFlags,
                         ret = FALSE;
                     for (i = 0; ret && i < msg->msg_data.info->cSignerInfo; i++)
                     {
-                        if (info->rgSigners[i].SignerId.dwIdChoice ==
-                         CERT_ID_KEY_IDENTIFIER)
+                        if (info->rgSigners[i].cbSize == sizeof(CMSG_SIGNER_ENCODE_INFO_WITH_CMS) &&
+                            info->rgSigners[i].SignerId.dwIdChoice == CERT_ID_KEY_IDENTIFIER)
                             msg->msg_data.info->version = CMSG_SIGNED_DATA_V3;
                         ret = CSignerInfo_Construct(
                          &msg->msg_data.info->rgSignerInfo[i],
-- 
2.47.1

